File: Stack/1798-max-number-of-k-sum-pairs/max-number-of-k-sum-pairs.java
class Solution {
    public int maxOperations(int[] nums, int k) {
        Map <Integer, Integer> map = new HashMap<>();
        int count = 0;
        for(int num: nums) {
            if(map.containsKey((k-num))) {
                count++;
                map.put(k-num, map.get(k-num)-1);
                if(map.get(k-num) == 0) {
                    map.remove(k-num);
                }
            } else {
                map.put(num, map.getOrDefault(num, 0) + 1);
            }
        }

        return count;
    }
}

File: Array/1798-max-number-of-k-sum-pairs/max-number-of-k-sum-pairs.py
class Solution:
    def maxOperations(self, nums: List[int], k: int) -> int:
        found_sum = defaultdict(int)
        count = 0
        for num in nums:
            if (k-num) in found_sum:
                count+=1
                found_sum[k-num]-=1
                if found_sum[k-num] == 0:
                    del found_sum[k-num]
            else:
                found_sum[num] += 1
        return count